ENH: Centralize fMRIPrep's and MRIQC's guidelines for Docker & DataLad (#47)
nipreps/nipreps.github.io: docs/apps/docker.md

Reviewer: On line 10, can we link to the Docker install instructions (https://docs.docker.com/engine/install/)?

because the user will not be able to action on the outputs otherwise.

Either way (whether the container is available with default settings
or the execution has been customized to normal users),
running as a normal user allows preempting these permissions issues.
This can be achieved with
[*Docker*'s `-u`/`--user` option](https://docs.docker.com/engine/containers/run/#user):

```
--user=[ user | user:group | uid | uid:gid | user:gid | uid:group ]
```

We can combine this option with *Bash*'s `id` command to ensure the current user's uid and group id (gid) are being set:

``` {.shell hl_lines="4"}
$ docker run -ti --rm \
    -v path/to/data:/data:ro \
    -v path/to/output:/out \
    -u $(id -u):$(id -g) \                   # set execution uid:gid
    -v path/to/tf-cache:/opt/templateflow \  # mount from host
    -e TEMPLATEFLOW_HOME=/opt/templateflow \ # override TF home
    nipreps/fmriprep:<latest-version> \
    /data /out/out \
    participant
```

For example:

``` Shell
$ docker run -ti --rm \
    -v $HOME/ds005:/data:ro \
    -v $HOME/ds005/derivatives:/out \
    -v $HOME/tmp/ds005-workdir:/work \
    -u $(id -u):$(id -g) \

>     -v $HOME/.cache/templateflow:/opt/templateflow \

Reviewer 2: mightn't it cause issues in case of parallel execution where multiple runs would compete over?

Author: I'll add an admonition about this. Thanks

Author: Added in 44360cb. Please send further suggestions or reopen this conversation if that fell short of addressing the issue.

If you need a finer control over the container execution, or you feel comfortable with the Docker Engine, avoiding the extra software layer of the wrapper might be a good decision.

### Accessing filesystems in the host within the container

> Containers are confined in a sandbox, so they can't access the host
> in any ways unless you explicitly prescribe acceptable accesses
> to the host.

Reviewer 3: This reads weirdly.

```suggestion
Containers are confined in a sandbox, so they can't access the data on the host
unless explicitly enabled.
```

> The Docker Engine provides mounting filesystems into the container with the `-v` argument and the following syntax:

Reviewer 3: ```suggestion
The Docker Engine supports mounting filesystems into the container
with the `-v` argument, using the syntax:
```

`-v some/path/in/host:/absolute/path/within/container:ro`,
where the trailing `:ro` specifies that the mount is read-only.
The mount permissions modifiers can be omitted, which means the mount
will have read-write permissions.

!!! warning "*Docker for Windows* requires enabling Shared Drives"

    On *Windows* installations, the `-v` argument will not work
    by default because it is necessary to enable shared drives.
    Please check on this [Stackoverflow post](https://stackoverflow.com/a/51822083) how to enable them.

In general, you'll want to at least provide two mount-points:
one set in read-only mode for the input data and one read/write
to store the outputs:

``` {.shell hl_lines="2 3"}
$ docker run -ti --rm \
    -v path/to/data:/data:ro \        # read-only, for data
    -v path/to/output:/out \          # read-write, for outputs
    nipreps/fmriprep:<latest-version> \
    /data /out/out \
    participant
```

> When **debugging** or **reusing pre-cached intermediate results**,
> you'll also need to mount some working directory that otherwise
> is not exposed by the application.
> In the case of *NiPreps*, we typically inform the *BIDS Apps*
> to override the work directory by setting the `-w`/`--work-dir`
> argument (please note that this is not defined by the *BIDS Apps*
> specifications and it may change across applications):

Reviewer 3: ```suggestion
We recommend mounting a work (or scratch) directory for intermediate workflow results.
This is particularly useful for **debugging** or **reusing pre-cached intermediate results**,
but can also be useful to control where these (large) directories get created,
as the default location for files created inside a docker container may not have sufficient space.
In the case of *NiPreps*, we typically inform the *BIDS Apps*
to override the work directory by setting the `-w`/`--work-dir`
argument (please note that this is not defined by the *BIDS Apps*
specifications and it may change across applications):
```

``` {.shell hl_lines="4 8"}
$ docker run -ti --rm \
    -v path/to/data:/data:ro \
    -v path/to/output:/out \
    -v path/to/work:/work \              # mount from host

>     nipreps/fmriprep:<latest-version> \

Reviewer 2: btw -- isn't there `freesurfer` license needed to be passed somehow?

Reviewer: Further up, on line 18, it says that all examples will refer to MRIQC (which doesn't require freesurfer license).

Author: Yes, I had the freesurfer license issue in mind. At the moment it's defined in fMRIPrep's documentation but we probably want to bubble it up here.

    /data /out/out \
    participant
    -w /work                             # override default directory
```

!!! tip "Best practices"

    The [*ReproNim* initiative](https://www.repronim.org/)
    distributes materials and documentation of best practices
    for containerized execution of neuroimaging workflows.
    Most of these are organized within the
    [*YODA* (Yoda's Organigram on Data Analysis)](https://github.com/myyoda) principles.

    For example, mounting `$PWD` into `$PWD` and setting that path
    as current working directory can effectively resolve many issues.
    This strategy may be combined with the above suggestion about
    the application's work directory as follows:

>     ``` {.shell hl_lines="4 5 9"}
>     $ docker run -ti --rm \
>         -v path/to/data:/data:ro \
>         -v path/to/output:/out \
>         -v $PWD:$PWD \
>         -w $PWD \   # DO NOT confuse with the application's work directory
>         nipreps/fmriprep:<latest-version> \
>         /data /out/out \
>         participant
>         -w $PWD/work
>     ```

Reviewer 3: I would personally do this:

```suggestion
    ``` {.shell hl_lines="4 5 9"}
    $ docker run -ti --rm \
        -v $PWD:$PWD \  # Mount the current directory with its own name
        -w $PWD \       # DO NOT confuse with the application's work directory
        nipreps/fmriprep:<latest-version> \
        inputs/raw-data outputs/fmriprep \  # With YODA, the inputs are inside the working directory
        participant
        -w $PWD/work
    ```
```

(Actually, I would use `sourcedata/raw .`, but that would require additional explanation.)

    Mounting `$PWD` may be used with YODA so that all necessary *parts*
    in execution are reachable from under `$PWD`.
    This effectively
    (i) makes it easy to *transfer* configurations from
    *outside* the container to the *inside* execution runtime;
    (ii) the *outside*/*inside* filesystem trees are homologous, which
    makes post-processing and orchestration easier;
    (iii) execution in shared systems is easier as everything is
    sort of *self-contained*.

    In addition to mounting `$PWD`, other advanced practices
    include mounting specific configuration files (for example, a
    [*Nipype* configuration file](https://miykael.github.io/nipype_tutorial/notebooks/basic_execution_configuration.html))
    into the appropriate paths within the container.

*BIDS Apps* relying on [*TemplateFlow*](https://www.templateflow.org)

> for atlases and templates management may require

Reviewer 3: ```suggestion
for atlas and templat management may require
```

the *TemplateFlow Archive* be mounted from the host.
Mounting the *Archive* from the host is an effective way

> to preempt the download of your favorite templates in every run:

Reviewer 3: ```suggestion
to prevent multiple downloads of templates that are not bundled in the image:
```

    -v path/to/tf-cache:/opt/templateflow \  # mount from host
    -e TEMPLATEFLOW_HOME=/opt/templateflow \ # override TF home
    nipreps/fmriprep:<latest-version> \
    /data /out/out \
    participant
    -w /work
```

!!! danger "Sharing the *TemplateFlow* cache can cause race conditions in parallel execution"

    When sharing the *TemplateFlow* *HOME* folder across several parallel
    executions against a single filesystem, these instance will likely
    attempt to fetch unavailable templates without sufficient time between
    actions for the data to be fully downloaded (in other words,
    data downloads will be *racing* each other).

    To resolve this issue, you will need to make sure all necessary
    templates are already downloaded within the cache folder.
    If the *TemplateFlow Client* is properly installed in your system,
    this is possible with the following command line
    (example shows how to fully download `MNI152NLin2009cAsym`:

    ``` Shell
    $ templateflow get MNI152NLin2009cAsym
    ```

### Running containers as a user

By default, Docker will run the container with the
user id (uid) **0**, which is reserved for the default **root**

> account in *Linux*.

Reviewer 2: Unless the actual docker there is `podman` which by default would run rootless ;)

```
ha@hopa:~$ rm -f 123; docker run -it -v $PWD:$PWD -w $PWD --entrypoint bash nipreps/fmriprep:latest -c 'echo `whoami` am in; id; pwd; touch 123; ls -l 123' ; echo "`whoami` is out"; ls -ld 123
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
root am in
uid=0(root) gid=0(root) groups=0(root)
/home/ha
-rw-r--r-- 1 root root 0 Jan  7 14:19 123
ha is out
-rw-r--r-- 1 ha ha 0 Jan  7 09:19 123
```

I am yet to try podman on hpc/for hpc, using primarily for services, but googled into  https://github.com/NERSC/podman-hpc ... worth researching and at least pointing users that there is another OCI solution podman which might give them easier means to run compute on their infrastructure.

In other words, by default *Docker* will use the superuser account
to execute the container and will write files with the corresponding
uid=0 unless configured otherwise.

> Executing as superuser may derive in permissions and security issues,

Reviewer 3: ```suggestion
Executing as superuser may result in permissions and security issues,
```

for example, [with *DataLad* (discussed later)](datalad.md#).
One paramount example of permissions issues where beginners typically
run into is deleting files after a containerized execution.
If the uid is not overridden, the outputs of a containerized execution
will be owned by **root** and group **root**.
Therefore, normal users will not be able to modify the output and

> superuser permissions will be required to deleted data generated

Reviewer 3: ```suggestion
superuser permissions will be required to delete data generated
```

by the containerized application.
Some shared systems only allow running containers as a normal user

> because the user will not be able to action on the outputs otherwise.

Reviewer 3: ```suggestion
because the user will not be able to operate on the outputs otherwise.
```

> Either way (whether the container is available with default settings
> or the execution has been customized to normal users),

Reviewer 3: ```suggestion
Whether the container is available with default settings,
or the execution has been customized to normal users,
```

> running as a normal user allows preempting these permissions issues.

Reviewer 3: ```suggestion
running as a normal user avoids these permissions issues.
```

``` {.shell hl_lines="4"}
$ docker run -ti --rm \
    -v path/to/data:/data:ro \
    -v path/to/output:/out \
    -u $(id -u):$(id -g) \                   # set execution uid:gid
    -v path/to/tf-cache:/opt/templateflow \  # mount from host
    -e TEMPLATEFLOW_HOME=/opt/templateflow \ # override TF home
    nipreps/fmriprep:<latest-version> \
    /data /out/out \
    participant
```

For example:

``` Shell
$ docker run -ti --rm \
    -v $HOME/ds005:/data:ro \
    -v $HOME/ds005/derivatives:/out \
    -v $HOME/tmp/ds005-workdir:/work \
    -u $(id -u):$(id -g) \
    -v $HOME/.cache/templateflow:/opt/templateflow \
    -e TEMPLATEFLOW_HOME=/opt/templateflow \
    nipreps/fmriprep:<latest-version> \
    /data /out/fmriprep-<latest-version> \
    participant \
    -w /work
```

### Application-specific options

Once the Docker Engine arguments are written, the remainder of the
command line follows the interface defined by the specific
*BIDS App* (for instance,
[*fMRIPrep*](https://fmriprep.readthedocs.io/en/latest/usage.html)
or [*MRIQC*](https://mriqc.readthedocs.io/en/latest/running.html#command-line-interface)).

> The first section of a call comprehends arguments specific to *Docker*,
> and configure the execution of the container:

Reviewer 3: ```suggestion
The first section of a call consists of arguments specific to *Docker*,
which configure the execution of the container:
```

    /data /out/fmriprep-<latest-version> \
    participant \
    -w /work
```

Finally, the application-specific options can be added.
We already described the work directory setting before, in the case
of *NiPreps* such as *MRIQC* and *fMRIPrep*.
Some options are *BIDS Apps* standard, such as
the *analysis level* (`participant` or `group`)
and specific participant identifier(s) (`--participant-label`):

``` {.shell hl_lines="9-12"}
$ docker run -ti --rm \
    -v $HOME/ds005:/data:ro \
    -v $HOME/ds005/derivatives:/out \
    -v $HOME/tmp/ds005-workdir:/work \
    -u $(id -u):$(id -g) \
    -v $HOME/.cache/templateflow:/opt/templateflow \
    -e TEMPLATEFLOW_HOME=/opt/templateflow \
    nipreps/fmriprep:<latest-version> \
    /data /out/fmriprep-<latest-version> \
    participant \
    --participant-label 001 002 \
    -w /work
```

### Resource constraints

*Docker* may be executed with limited resources.
Please [read the documentation](https://docs.docker.com/engine/containers/resource_constraints/)
to limit resources such as memory, memory policies, number of CPUs, etc.

**Memory will be a common culprit** when working with large datasets

> (+10GB).

Reviewer 2: another shameless plug which might be of interest/help.

Inspired by our reproman, and BrainLife's helper to monitor execution of compute, we created a simple helper https://github.com/con/duct which could be of help to monitor/identify memory and cpu requirements for e.g. future informed specification for job parameters or plotting resource consumption during compute. It also takes care about storing stdout/stderr outputs produced, thus making it possible (if used along with `datalad *run`) to capture those for possible future troubleshooting etc.

Author: If you finally send an additional PR, I'm happy to see this documented there :)
